docs/how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,124 @@
+# How does version resolution work in Maven and Gradle?
+
+The build system that you choose for building your Java code determines which
+version resolution algorithm is used to choose the versions of your
+dependencies. Unfortunately, the two most popular build systems in open source
+(Maven and Gradle) use different version resolution algorithms. They also use
+different names for the concept: Maven calls the process *dependency mediation*,
+while Gradle calls the process *version conflict resolution*. In this article,
+we will use the term *version resolution*, omitting the word "conflict" so that
+the process it not confused with program-breaking conflicts that we discuss
+elsewhere. No matter which build system you choose, you are likely to have
+direct or indirect dependencies that use the other one, so you need to
+understand how both work. Note: Our focus is on build systems based on the Maven
+repository system, so we ignore other build environments like Eclipse and Ant
+that work differently.
+
+Before we look at Maven and Gradle specifically, we need to understand what
+happens in the version resolution process. The input of this process is a
+dependency graph which reflects the dependencies as specified by each library in
+the graph, which we call an *unresolved dependency graph*. In such a graph, there
+can be multiple versions of each library. The version resolution process walks
+the unresolved dependency graph and decides which version to use for every
+library encountered. Whether or not the build tool produces a graph as an output
+of this resolution, it is useful to think of this output in terms of a new
+graph. We call the graph representation of this output a *resolved dependency
+graph*. Here is a visual representation of the process:
+
+<img src="assets/images/ddc-resolution-00.png">
+
+For example, we could have an unresolved graph that looks like this:
+
+<img src="assets/images/ddc-resolution-01.png">
+
+In this graph, there are two versions of A (10.0 and 10.1). The version
+resolution process picks a single version for A, which could be either version
+depending on the algorithm. If A:10.0 is picked, the resolved dependency graph
+looks like the following. In this graph, C's dependency on A:10.1 is overridden.
+
+<img src="assets/images/ddc-resolution-02.png">
+
+If A:10.1 is picked, the resolved dependency graph looks like the following. In
+this graph, B's dependency on A:10.0 is overridden.
+
+<img src="assets/images/ddc-resolution-03.png">
+
+So why do we need to pick a single version? The reason is that Java's default
+class loader only loads one version of each fully-qualified class name at
+runtime. So, if you try to put multiple versions on the classpath, you would not
+get both versions at runtime. Java build systems handle this issue upfront by
+picking a single version for each library before the classpath is even
+constructed.
+
+Now let's look at the version resolution algorithms used by Gradle and
```

Comment:
Google tech writer style guide discourages "Let's'" and recommends second and third person as appropriate: 

https://developers.google.com/style/person